### PR TITLE
Configurator Baseline Code

### PR DESCRIPTION
**Pull request by garonsky**
```
- ConfigMgr 2.0 Initial baseline
```

Signed-off-by: Gleb Aronsky gleb.aronsky@lexisnexis.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
There are a couple of files called .CMakeLists.txt.swp included that probably should not be.

There is also a file configuration/configurator/dump.txt (empty) that should be removed.

A number of the *.cpp and *.h* files also appear to be empty.

Is there a reason why you default the building of this component to off?

**Comment by garonsky** [comment]
@richardkchapman 
I have removed temp and empty files from the pull request.
The command line based configurator will be built by default.

File: configuration/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+################################################################################
+
+
+# Check if environment variable QTDIR is set.
+IF( CONFIGURATOR_QT_UI AND DEFINED ENV{QTDIR} )
+    SET( CMAKE_PREFIX_PATH $ENV{QTDIR} ${CMAKE_PREFIX_PATH} )
```
Comment (richardkchapman) [inline configuration/CMakeLists.txt]: Shouldn't this use a standard FindQT mechanism?

**Comment by richardkchapman** [comment]
There is still a file configuration/configurator/.CMakeLists.txt.swp in there that should not be

File: configuration/configurator/ConfigSchemaHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+{
+    //assert(comp != NULL && *comp != 0);
+    if (! (comp != NULL && *comp != 0) )
+    {
+        DBGLOG("no component selected for QML, index = %d", nIdx);
+        return NULL;
+    }
+    assert(m_pSchemaMapManager != NULL);
+
+    static StringBuffer strQML;
```
Comment (richardkchapman) [inline configuration/configurator/ConfigSchemaHelper.cpp]: Why static? Other than making sure it's not threadsafe, I see no benefit...

File: configuration/configurator/ConfigSchemaHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+        if (pSchemaName != NULL && strcmp(comp, pSchemaName) == 0)
+        {
+             pSchema = m_pSchemaMapManager->getSchemaForXSD(pSchemaName);
+
+             assert(pSchema != NULL);
+
+             if (pSchema != NULL)
+             {
+                 pSchema->getQML3(strQML, nIdx);
+                 return strQML.str();
```
Comment (richardkchapman) [inline configuration/configurator/ConfigSchemaHelper.cpp]: Ok, I see why it is static. But it's a terrible idea. Pass in a StringBuffer reference.

File: configuration/configurator/ConfigSchemaHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+             {
+                pSchema->dump(std::cout);
+            }
+        }
+    }
+}
+
+void CConfigSchemaHelper::dumpStdOut() const
+{
+    assert("NOT IMPLEMENTED");
```
Comment (richardkchapman) [inline configuration/configurator/ConfigSchemaHelper.cpp]: Is it something that needs to be implemented some time? In which case the UNIMPLEMENTED macro would be appropriate. Or is it something that should never be executed (e.g. overridden in derived classes)? In which case throwUnexpected() would be appropriate.

File: configuration/configurator/ConfigSchemaHelper.cpp [hunk 5]
```diff
@@ -0,0 +1,889 @@
...
+
+void CConfigSchemaHelper::processExtensionArr()
+{
+    int length = m_extensionArr.length();
+
+    for (int idx = 0; idx < length; idx++)
+    {
+        CExtension &Extension = (m_extensionArr.item(idx));
+        const char *pName = Extension.getBase();
+
```
Comment (richardkchapman) [inline configuration/configurator/ConfigSchemaHelper.cpp]: minor: An awful lot of blank lines in here

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+
+    return arrComps.length();*/
+    return CBuildSetManager::getInstance()->getBuildSetComponentCount();
+}
+
+int getNumberOfAvailableServices()
+{
+    assert(s_pConfigSchemaHelper != NULL);
+
+    /*StringArray arrServices;
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: There's quite a bit of commented out code in here. Is it serving any purpose?

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+
+            //return pAttribute->getEnvValueFromXML();
+            return pAttribute->getInstanceValue();
+        }
+    }
+}
+
+void setTableValue(const char *pXPath, int index, const char *pValue)
+{
+    assert(false);  // NOT IMPLEMENTED
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: See previous comments about UNIMPLEMENTED/throwUnexpected()

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+    else
+    {
+        assert( compIdx >= 0);
+
+        CElementArray *pElementArray = static_cast<CElementArray*>(pData);
+
+        assert(pElementArray->getNodeType() == XSD_ELEMENT_ARRAY);
+
+        const CXSDNodeBase *pNodeBase = &(pElementArray->item(compIdx+idx));
+
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: Are you being paid by the line? Why so many blank lines? They make it a lot harder to review the code (GitHub is struggling with the size of the change as it is).

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+int getIndexFromParent(void *pData)
+{
+ //   assert(pData != NULL);
+
+    CEnvironmentModelNode *pNode = static_cast<CEnvironmentModelNode*>(pData);
+
+    //assert (pNode->getParent() != NULL);
+    if (pNode->getParent() == NULL)
+    {
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: Slightly overboard on the number of curly braces too...

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 10]
```diff
@@ -0,0 +1,745 @@
...
+{
+    if (pXSD == NULL || *pXSD == 0)
+    {
+        return NULL;
+    }
+
+    //const char *pFileName = ConfigSchemaHelper::getInstance()->getFileName(pData)
+
+    //return CConfigSchemaHelper::getInstance()->printQML(pFileName, 0);
+            return NULL;
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: Weird indentation (and commented-out code)

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+    bool bRetVal = false;
+
+    bRetVal = CConfigSchemaHelper::getInstance()->saveConfigurationFile();
+
+    return bRetVal;
+}
+
+int getNumberOfNotificationTypes()
+{
+    int nRetVal = 0;
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: Why not just "return CNotificationManager:: ..."? 5 lines where one would do.

File: configuration/configurator/ConfiguratorAPI.hpp [hunk 12]
```diff
@@ -0,0 +1,193 @@
+#ifndef _CONFIGURATOR_API_HPP_
+#define _CONFIGURATOR_API_HPP_
+
+#define MAX_ARRAY_X 100
+#define MAX_ARRAY_Y 128
+
+const char modelNames[MAX_ARRAY_X][MAX_ARRAY_Y] =
+                                        {   "tableDataModel0",
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.hpp]: This does not feel like the right way to do whatever it is you are trying to do here...

File: configuration/configurator/QMLMarkup.cpp [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    StringBuffer result;
+    StringBuffer escapedToolTip = tooltip;
+    escapedToolTip.replaceString("\"","\\\"");
+    StringBuffer escapedPlaceholder = placeholder;
+    escapedPlaceholder.replaceString("\"","\\\"");
+    result.append(role);
+    result.append(": ListElement { \n");
+    result.append("value:[");
+    for(int i = 0; i < values.length(); i++)
+{        result.append("ListElement {value: \"");
```
Comment (richardkchapman) [inline configuration/configurator/QMLMarkup.cpp]: Formatting gone odd

File: configuration/configurator_app/configurator_app.creator [hunk 14]
```diff
@@ -0,0 +1 @@
+[General]
```
Comment (richardkchapman) [inline configuration/configurator_app/configurator_app.creator]: Should this have been checked in?

File: configuration/configurator_app/configurator_app.files [hunk 15]
```diff
@@ -0,0 +1,14 @@
+Main.cpp
```
Comment (richardkchapman) [inline configuration/configurator_app/configurator_app.files]: Should this have been checked in?

**Comment by richardkchapman** [comment]
This change is a bit too large for GitHub's file view making it hard for me to review and comment on. So far I've really only been able to look at pretty general style issues, but even commenting on them is painfully slow.

I wonder if it would make sense to split into several smaller commits - perhaps one for each subtree - that could be reviewed/commented on separately?

**Comment by richardkchapman** [comment]
@garonsky Back to you for now

**Pull request closed without merging**